### PR TITLE
Update the MOI test file to use caching optimizer

### PR DESCRIPTION
**Pull request by odow**
Let's see if these changes help more tests to pass.

### CODE REVIEW

File: test/testMOI.jl [hunk 1]
```diff
@@ -88,40 +13,29 @@ function runtests()
...
-Pass arguments to `exclude` to skip tests for functionality that is not
-implemented or that your solver doesn't support.
-"""
 function test_runtests()
-    MOI.Test.runtests(
-        BRIDGED,
-        CONFIG,
-        exclude=[
-            "test_attribute_NumberOfThreads",
-            "test_quadratic_",
```
Comment (odow) [inline test/testMOI.jl:138]: This, for example, skipped all quadratic tests, so I don't think it's what you intended :smile:

**Comment by odow** [comment]
@PharosAbad you just need to click "approve" for the CI to start running.

**Comment by odow** [comment]
Looks like many tests fail with `PosDefException: matrix is not positive definite; Cholesky factorization failed.`. 

You're missing a few attributes, `PrimalStatus`, `DualStatus`, `RawStatusString`
https://github.com/PharosAbad/StatusSwitchingQP.jl/blob/baa9204c2dec211c1ff3239e7142f5f73755d62c/src/MOIwrapper.jl#L145-L147

`test_linear_open_intervals` fails with the wrong solution

You can ignore these three for now:

`test_model_LowerBoundAlreadySet`
`test_model_UpperBoundAlreadySet`
`test_model_supports_constraint_ScalarAffineFunction_EqualTo`


You can run each example as
```julia
model = MOI.instantiate(StatusswitchingQP.Optimizer; with_cache_type = Float64, with_bridge_type = Float64)
config = MOI.Test.Config()
MOI.Test. test_linear_open_intervals(model, config)
```

**Comment by PharosAbad** [comment]
the [patch-1](https://github.com/odow/StatusSwitchingQP.jl/tree/patch-1) branch : Some tests did not pass: 867 passed, 7 failed, 140 errored, 0 broken.

what to do then?

**Comment by PharosAbad** [comment]
for the `test_linear_open_intervals`, my solver convert it to  a pair of inequalities constraints

`d <= ax+b <= u` to   `ax+b <= u` and ` -ax-b <= -d`

```
┌ Warning: repacking AffineFunction range rows: 
│   s.lower = -Inf
│   s.upper = 1.0
└ @ StatusSwitchingQP /3T/TeXLive/StatusSwitchingQP/src/MOIwrapper.jl:279
┌ Warning: repacking AffineFunction range rows: 
│   s.lower = -1.0
│   s.upper = Inf
└ @ StatusSwitchingQP /3T/TeXLive/StatusSwitchingQP/src/MOIwrapper.jl:279
┌ Warning: repacking AffineFunction range rows: 
│   s.lower = -Inf
│   s.upper = Inf
└ @ StatusSwitchingQP /3T/TeXLive/StatusSwitchingQP/src/MOIwrapper.jl:279
Test Failed at /home/dengta/.julia/packages/MathOptInterface/goW8i/src/Test/test_linear.jl:4049
  Expression: ≈(MOI.get(model, MOI.VariablePrimal(), x), T[1, -1, 1], config)
   Evaluated: ≈([-1.0, 0.0, 0.0], [1.0, -1.0, 1.0], MathOptInterface.Test.Config{Float64}(1.4901161193847656e-8, 1.4901161193847656e-8, MathOptInterface.OPTIMAL, MathOptInterface.INFEASIBLE, Any[]))
```

**Comment by odow** [comment]
> my solver convert it to a pair of inequalities constraints

Sure. But if it returns the wrong answer, then there must be a bug somewhere.

You can also let JuMP do that for you. Remove `MOI.Interval` from
https://github.com/PharosAbad/StatusSwitchingQP.jl/blob/baa9204c2dec211c1ff3239e7142f5f73755d62c/src/MOIwrapper.jl#L122
and remove
https://github.com/PharosAbad/StatusSwitchingQP.jl/blob/baa9204c2dec211c1ff3239e7142f5f73755d62c/src/MOIwrapper.jl#L269-L280

In general, you should only add support to MOI for constraints that your solver explicitly supports.

**Comment by maxchendt** [comment]
> You can also let JuMP do that for you. Remove `MOI.Interval` from
> and remove
> In general, you should only add support to MOI for constraints that your solver explicitly supports.

no luck, after the removement

```
Test Failed at  .julia/packages/MathOptInterface/goW8i/src/Test/test_linear.jl:4049
  Expression: ≈(MOI.get(model, MOI.VariablePrimal(), x), T[1, -1, 1], config)
   Evaluated: ≈([-0.0, -0.0, 0.0], [1.0, -1.0, 1.0], MathOptInterface.Test.Config{Float64}(1.4901161193847656e-8, 1.4901161193847656e-8, MathOptInterface.OPTIMAL, MathOptInterface.INFEASIBLE, Any[]))
```

**Comment by odow** [comment]
Per https://github.com/jump-dev/JuMP.jl/pull/3354#issuecomment-1551546300, you don't really need to know the mapping. Just solve the problem as given, and JuMP will do the rest.

If you're getting the wrong solution, then there are bugs somewhere

**Comment by PharosAbad** [comment]
> Looks like many tests fail with `PosDefException: matrix is not positive definite; Cholesky factorization failed.`.


my program: make sure that  `J > 0 || any(isfinite.(d)) || any(isfinite.(u)) `. Otherwise, there are no any inequalities or bounds, there are only equality constraints. For QP, there is a known closed-form solution, and for LP, the solution is unbounded.

and some of the test data from MOI runtest  have only inequalities constraints.  hence, there are many errors during the test.


Furthermore, for the objective function,  QP and LP use different solver. QP requires that MOI.ObjectiveFunction{MOI.ScalarQuadraticFunction{T}} has a positive semi-definite matrix but not zero. and QP use LP to find the starting point.

**Comment by odow** [comment]
The tests are there so that every solver interfaced to JuMP performs somewhat similarly. 

Take `test_attribute_RawStatusString `

image

The test is
https://github.com/jump-dev/MathOptInterface.jl/blob/197424ddf1e03ccabc3d9b099f0f86dca8165752/src/Test/test_attribute.jl#L43-L51
which is basically
```julia
MOI.add_variable(model)
MOI.optimize!(model)
@test MOI.get(model, MOI.RawStatusString()) isa String
```

So instead of throwing the assertion error, what you should do is return a status like `"no inequalities or variable bounds"`, and `MOI.TerminationStatus` should be `MOI.INVALID_MODEL`.

Of course, a better return would be to return `x = 0` because if there are no constraints, bounds, or objective, then any point is optimal.


For the dual status failures, this line is incorrect: 
https://github.com/PharosAbad/StatusSwitchingQP.jl/blob/642bd04e815b9edfd702a8aa4d9d41d0b3ec100c/src/MOIwrapper.jl#L254

Every optimizer _must_ support DualStatus. The correct method is `MOI.get(::Optimizer, ::DualStatus) = MOI.NO_SOLUTION`.

**Comment by odow** [comment]
> For QP, there is a known closed-form solution, and for LP, the solution is unbounded.

So the correct thing here is to return the QP solution, and the return a `TerminationStatus` of `MOI.DUAL_INFEASIBLE` for the LP.

**Comment by PharosAbad** [comment]
> `test_linear_open_intervals` fails with the wrong solution

```
Test Failed at /home/dengta/.julia/packages/MathOptInterface/goW8i/src/Test/test_linear.jl:4049
  Expression: ≈(MOI.get(model, MOI.VariablePrimal(), x), T[1, -1, 1], config)
   Evaluated: ≈([1.0, 1.0, -1.0], [1.0, -1.0, 1.0], MathOptInterface.Test.Config{Float64}(1.4901161193847656e-8, 1.4901161193847656e-8, MathOptInterface.OPTIMAL, MathOptInterface.INFEASIBLE, Any[]))
```

we can see

```Julia
import MathOptInterface as MOI
import MathOptInterface.Utilities as MOIU
T = Float64
model = MOIU.Model{T}()
x = MOI.add_variables(model, 3)
MOI.set(model, MOI.ObjectiveSense(), MOI.MAX_SENSE)
f = T(1) * x[1] + T(-1) * x[2] + T(1) * x[3]
MOI.set(model, MOI.ObjectiveFunction{typeof(f)}(), f)
MOI.add_constraint(model, T(1) * x[1], MOI.Interval(typemin(T), T(1)))
MOI.add_constraint(model, T(1) * x[2], MOI.Interval(T(-1), typemax(T)))
MOI.add_constraint(model, T(1) * x[3], MOI.Interval(typemin(T), typemax(T)))
MOI.add_constraint(model, x[3], MOI.LessThan(T(1)))

using StatusSwitchingQP, JuMP

O = JuMP.Model(StatusSwitchingQP.Optimizer)
ix = MOI.copy_to(O, model)
display(ix)
```

shows that

```
MathOptInterface.Utilities.IndexMap with 7 entries:
  MOI.VariableIndex(3)                                                 => MOI.VariableIndex(1)
  MOI.VariableIndex(1)                                                 => MOI.VariableIndex(2)
  MOI.VariableIndex(2)                                                 => MOI.VariableIndex(3)
  ConstraintIndex{VariableIndex, LessThan{Float64}}(3)                 => ConstraintIndex{VariableIndex, LessThan{Float64}}(1)
  ConstraintIndex{ScalarAffineFunction{Float64}, Interval{Float64}}(2) => ConstraintIndex{ScalarAffineFunction{Float64}, Interval{Float64}}(2)
  ConstraintIndex{ScalarAffineFunction{Float64}, Interval{Float64}}(3) => ConstraintIndex{ScalarAffineFunction{Float64}, Interval{Float64}}(3)
  ConstraintIndex{ScalarAffineFunction{Float64}, Interval{Float64}}(1) => ConstraintIndex{ScalarAffineFunction{Float64}, Interval{Float64}}(1)
```

Hence, ` [1.0, 1.0, -1.0]`  is map to ` [1.0, -1.0, 1.0]`  , as index 3->1, 1->2 and 2 ->3.   I do not know how to fix this

**Comment by odow** [comment]
Presumably there's a bug here:
https://github.com/PharosAbad/StatusSwitchingQP.jl/blob/642bd04e815b9edfd702a8aa4d9d41d0b3ec100c/src/MOIwrapper.jl#L175-L194

This index map needs to map the variables from `src` to the variables that you've defined in `dest`:
https://github.com/PharosAbad/StatusSwitchingQP.jl/blob/642bd04e815b9edfd702a8aa4d9d41d0b3ec100c/src/MOIwrapper.jl#L163

**Comment by PharosAbad** [comment]
> Presumably there's a bug here

fix https://github.com/PharosAbad/StatusSwitchingQP.jl/commit/bc13f343fd92d49655adaef5cdd6bf6ba9d764ea

**Comment by PharosAbad** [comment]
finally https://github.com/PharosAbad/StatusSwitchingQP.jl/commit/b62b22cbde32fe3c9e84876fef1d31609702fc66 , I hope to exclude the following

```
"test_constraint_ScalarAffineFunction_EqualTo",
"test_modification_const_scalar_objective",
"test_modification_const_vectoraffine_zeros",
"test_objective_ObjectiveFunction",
"test_objective_qp_ObjectiveFunction_edge_cases",
"test_quadratic_nonhomogeneous",
"test_solve_TerminationStatus_DUAL_INFEASIBLE",
"test_solve_result_index",
```

For I can not provide any information on dual problem.

This [benchmark instances](http://plato.asu.edu/ftp/milp.html) show that even the top solver can not solve all the problems tested.

**Comment by odow** [comment]
If the tests are failing, it's because there are bugs in your implementation. These are not numerically difficult tests to pass, so you should be able to pass 100% of the tests.

Here's an example of the first thing that came to mind:
```Julia
(base) oscar@Oscars-MBP StatusSwitchingQP % julia +1.9 --project=/tmp
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.9.0 (2023-05-07)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

(tmp) pkg> st
Status `/private/tmp/Project.toml`
  [4076af6c] JuMP v1.11.0
  [73a7d86c] StatusSwitchingQP v0.2.4 `https://github.com/PharosAbad/StatusSwitchingQP.jl.git#main`

julia> using JuMP

julia> import StatusSwitchingQP

julia> model = Model(StatusSwitchingQP.Optimizer)
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: StatusSwitchingQP

julia> @variable(model, x)
x

julia> @objective(model, Min, x^2 + 1)
x² + 1

julia> optimize!(model)
┌ Warning: no inequalities and bounds
└ @ StatusSwitchingQP ~/.julia/packages/StatusSwitchingQP/nYTVC/src/types.jl:250
ERROR: UndefVarError: `N` not defined
Stacktrace:
  [1] solveQP(Q::StatusSwitchingQP.QP{Float64}; settings::StatusSwitchingQP.Settings{Float64}, settingsLP::StatusSwitchingQP.Settings{Float64})
    @ StatusSwitchingQP.SSQP ~/.julia/packages/StatusSwitchingQP/nYTVC/src/SSQP.jl:227
  [2] solveQP
    @ ~/.julia/packages/StatusSwitchingQP/nYTVC/src/SSQP.jl:224 [inlined]
  [3] optimize!(opt::StatusSwitchingQP.Optimizer{Float64})
    @ StatusSwitchingQP ~/.julia/packages/StatusSwitchingQP/nYTVC/src/MOIwrapper.jl:220
  [4] optimize!
    @ ~/.julia/packages/MathOptInterface/goW8i/src/MathOptInterface.jl:85 [inlined]
  [5] optimize!(m::MathOptInterface.Utilities.CachingOptimizer{StatusSwitchingQP.Optimizer{Float64}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}})
    @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/goW8i/src/Utilities/cachingoptimizer.jl:316
  [6] optimize!
    @ ~/.julia/packages/MathOptInterface/goW8i/src/Bridges/bridge_optimizer.jl:376 [inlined]
  [7] optimize!
    @ ~/.julia/packages/MathOptInterface/goW8i/src/MathOptInterface.jl:85 [inlined]
  [8] optimize!(m::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{StatusSwitchingQP.Optimizer{Float64}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}})
    @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/goW8i/src/Utilities/cachingoptimizer.jl:316
  [9] optimize!(model::Model; ignore_optimize_hook::Bool, _differentiation_backend::MathOptInterface.Nonlinear.SparseReverseMode, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JuMP ~/.julia/packages/JuMP/AKvOr/src/optimizer_interface.jl:440
 [10] optimize!(model::Model)
    @ JuMP ~/.julia/packages/JuMP/AKvOr/src/optimizer_interface.jl:410
 [11] top-level scope
    @ REPL[7]:1
```

**Pull request closed without merging**